crates/sound/src/audio_backends/alsa.rs

```diff
@@ -0,0 +1,550 @@
+                virtio_sound::VIRTIO_SND_PCM_FMT_U20 => Format::u20_3(),
+                virtio_sound::VIRTIO_SND_PCM_FMT_S24 => Format::s24(),
+                virtio_sound::VIRTIO_SND_PCM_FMT_U24 => Format::u24(),
+                virtio_sound::VIRTIO_SND_PCM_FMT_S32 => Format::s32(),
+                virtio_sound::VIRTIO_SND_PCM_FMT_U32 => Format::r#u32(),
+                virtio_sound::VIRTIO_SND_PCM_FMT_FLOAT => Format::float(),
+                virtio_sound::VIRTIO_SND_PCM_FMT_FLOAT64 => Format::float64(),
+                virtio_sound::VIRTIO_SND_PCM_FMT_DSD_U8 => Format::DSDU8,
+                virtio_sound::VIRTIO_SND_PCM_FMT_DSD_U16 => Format::DSDU16LE,
+                virtio_sound::VIRTIO_SND_PCM_FMT_DSD_U32 => Format::DSDU32LE,
+                virtio_sound::VIRTIO_SND_PCM_FMT_IEC958_SUBFRAME => Format::iec958_subframe(),
+                _ => Format::Unknown,
+            })?;
+
+            hwp.set_access(Access::RWInterleaved)?;
+
+            // > A period is the number of frames in between each hardware interrupt.
+            // - https://www.alsa-project.org/wiki/FramesPeriods
+
+            // FIXME
+            if let Err(err) = hwp.set_buffer_size(4096) {
+                log::error!("Could not set buffer size of 4096 bytes: {}.", err,);
+            }
+            pcm.hw_params(&hwp)?;
+        }
+
+        // FIXME
+        //if let Err(err) = hwp.set_period_size(_, ValueOr::Greater) {
+        //        log::error!(
+        //            "Could not set period size: {}",
+        //            err,
+        //        );
+        //    }
+        //}
+
+        /*
+        //  {
+        // Make sure we don't start the stream too early
+        let hwp = pcm.hw_params_current()?;
+        let swp = pcm.sw_params_current()?;
+        //swp.set_start_threshold(hwp.get_buffer_size()?)?;
+        //swp.set_stop_threshold(hwp.get_buffer_size()?)?;
+        pcm.sw_params(&swp)?;
+        }
+        */
+        pcm
+    };
+    Ok(())
+}
+
+fn write_samples_direct(
+    pcm: &alsa::PCM,
+    stream: &mut Stream,
+    mmap: &mut alsa::direct::pcm::MmapPlayback<u8>,
+) -> AResult<bool> {
+    while mmap.avail() > 0 {
+        // Write samples to DMA area from iterator
+        //
+        let Some(buffer) = stream.buffers.front_mut() else {
+            return Ok(false);
+        };
+        let mut iter = buffer.bytes[buffer.pos..].iter().cloned();
+        let frames = mmap.write(&mut iter);
+        let written_bytes = pcm.frames_to_bytes(frames);
+        if let Ok(written_bytes) = usize::try_from(written_bytes) {
+            buffer.pos += written_bytes;
+        }
+        if buffer.pos >= buffer.bytes.len() {
+            stream.buffers.pop_front();
+        }
+    }
+    match mmap.status().state() {
+        State::Running => {
+            return Ok(false);
+        } // All fine
+        State::Prepared => {
+            //log::trace!("Starting audio output stream");
+            //p.start()?
```

Comment:
Is it expected the Prepared/Suspended parts are commented?

Reply:
They were left over from the ALSA crate's example code, removed